File: Final Round/Models/InterviewModels.swift
import Foundation

struct InterviewQuestion: Identifiable, Codable {
    let id: UUID
    let text: String
    let category: QuestionCategory
    let difficulty: Difficulty
    var answer: QuestionAnswer?
    
    init(id: UUID = UUID(), text: String, category: QuestionCategory, difficulty: Difficulty, answer: QuestionAnswer? = nil) {
        self.id = id
        self.text = text
        self.category = category
        self.difficulty = difficulty
        self.answer = answer
    }
}

struct QuestionAnswer: Codable {
    let transcription: String
    let audioURL: String?
    let videoURL: String?
    let evaluation: AnswerEvaluation?
    let timestamp: Date
    let eyeContactMetrics: EyeContactMetrics?
    let confidenceScore: Double?
    let toneAnalysis: ToneAnalysis?
    let timeSpent: TimeInterval?
    
    init(transcription: String, audioURL: String? = nil, videoURL: String? = nil, evaluation: AnswerEvaluation? = nil, timestamp: Date = Date(), eyeContactMetrics: EyeContactMetrics? = nil, confidenceScore: Double? = nil, toneAnalysis: ToneAnalysis? = nil, timeSpent: TimeInterval? = nil) {
        self.transcription = transcription
        self.audioURL = audioURL
        self.videoURL = videoURL
        self.evaluation = evaluation
        self.timestamp = timestamp
        self.eyeContactMetrics = eyeContactMetrics
        self.confidenceScore = confidenceScore
        self.toneAnalysis = toneAnalysis
        self.timeSpent = timeSpent
    }
}

// MARK: - Analytics Data Structures

struct EyeContactMetrics: Codable {
    let percentage: Double // 0-100
    let totalDuration: TimeInterval
    let lookingAtCameraDuration: TimeInterval
    let timestamps: [EyeContactTimestamp]
    
    var formattedPercentage: String {
        return String(format: "%.1f%%", percentage)
    }
}

struct EyeContactTimestamp: Codable {
    let time: TimeInterval
    let isLookingAtCamera: Bool
}

struct ToneAnalysis: Codable {
    let speechPace: Double // words per minute
    let pauseCount: Int
    let averagePauseDuration: TimeInterval
    let volumeVariation: Double // 0-1 scale
    let sentiment: SentimentScore
    
    var paceDescription: String {
        switch speechPace {
        case 0..<100: return "Slow"
        case 100..<140: return "Moderate"
        case 140..<180: return "Fast"
        default: return "Very Fast"
        }
    }
    
    var formattedPace: String {
        return String(format: "%.0f WPM", speechPace)
    }
}

struct SentimentScore: Codable {
    let score: Double // -1 (negative) to 1 (positive)
    let confidence: Double // 0-1
    
    var label: String {
        switch score {
        case 0.5...1.0: return "Very Positive"
        case 0.1..<0.5: return "Positive"
        case -0.1...0.1: return "Neutral"
        case -0.5..<(-0.1): return "Negative"
        default: return "Very Negative"
        }
    }
}

struct AnswerEvaluation: Codable {
    let score: Int
    let strengths: [String]
    let improvements: [String]
    let feedback: String
    
    var grade: String {
        switch score {
        case 95...100: return "A+"
        case 90..<95: return "A"
        case 85..<90: return "A-"
        case 80..<85: return "B+"
        case 75..<80: return "B"
        case 70..<75: return "B-"
        case 65..<70: return "C+"
        case 60..<65: return "C"
        case 55..<60: return "C-"
        default: return "D"
        }
    }
    
    var gradeColor: String {
        switch score {
        case 90...100: return "green"
        case 80..<90: return "blue"
        case 70..<80: return "orange"
        default: return "red"
        }
    }
}

enum QuestionCategory: String, CaseIterable, Codable {
    case behavioral = "Behavioral"
    case technical = "Technical"
    case situational = "Situational"
    case general = "General"
}

enum Difficulty: String, CaseIterable, Codable {
    case easy = "Easy"
    case medium = "Medium"
    case hard = "Hard"
}

struct InterviewSession: Identifiable, Codable {
    let id: UUID
    let role: String
    let difficulty: Difficulty
    let categories: [QuestionCategory]
    let duration: Int // in minutes
    var questions: [InterviewQuestion]
    var answeredCount: Int
    var skippedCount: Int
    var startTime: Date?
    var endTime: Date?
    var userEmail: String?
    var enableAudioRecording: Bool
    var experienceLevel: String // User's experience level for evaluation context
    
    init(id: UUID = UUID(), role: String, difficulty: Difficulty, categories: [QuestionCategory], duration: Int = 0, questions: [InterviewQuestion], answeredCount: Int = 0, skippedCount: Int = 0, startTime: Date? = nil, endTime: Date? = nil, userEmail: String? = nil, enableAudioRecording: Bool = true, experienceLevel: String = "Mid Level") {
        self.id = id
        self.role = role
        self.difficulty = difficulty
        self.categories = categories
        self.duration = duration // Legacy field, actual duration computed from startTime/endTime
        self.questions = questions
        self.answeredCount = answeredCount
        self.skippedCount = skippedCount
        self.startTime = startTime
        self.endTime = endTime
        self.userEmail = userEmail
        self.enableAudioRecording = enableAudioRecording
        self.experienceLevel = experienceLevel
    }
    
    var totalQuestions: Int {
        questions.count
    }
    
    var attemptedQuestions: Int {
        answeredCount + skippedCount
    }
    
    var remainingQuestions: Int {
        totalQuestions - attemptedQuestions
    }
    
    var sessionDuration: TimeInterval? {
        guard let start = startTime, let end = endTime else { return nil }
        return end.timeIntervalSince(start)
    }
    
    var completionRate: Double {
        guard totalQuestions > 0 else { return 0 }
        return Double(answeredCount) / Double(totalQuestions)
    }
    
    var averageScore: Double {
        let evaluatedQuestions = questions.compactMap { $0.answer?.evaluation }
        guard !evaluatedQuestions.isEmpty else { return 0 }
        let totalScore = evaluatedQuestions.reduce(0) { $0 + $1.score }
        return Double(totalScore) / Double(evaluatedQuestions.count)
    }
    
    var overallGrade: String {
        let score = Int(averageScore)
        switch score {
        case 95...100: return "A+"
        case 90..<95: return "A"
        case 85..<90: return "A-"
        case 80..<85: return "B+"
        case 75..<80: return "B"
        case 70..<75: return "B-"
        case 65..<70: return "C+"
        case 60..<65: return "C"
        case 55..<60: return "C-"
        default: return "D"
        }
    }
    
    var gradeColor: String {
        let score = Int(averageScore)
        switch score {
        case 90...100: return "green"
        case 80..<90: return "blue"
        case 70..<80: return "orange"
        default: return "red"
        }
    }
    
    var answerRate: Double {
        guard attemptedQuestions > 0 else { return 0 }
        return Double(answeredCount) / Double(attemptedQuestions)
    }
    
    var formattedDuration: String {
        guard let duration = sessionDuration else { return "N/A" }
        let minutes = Int(duration) / 60
        let seconds = Int(duration) % 60
        return "\(minutes)m \(seconds)s"
    }
    
    var formattedDate: String {
        guard let start = startTime else { return "N/A" }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .full
        return formatter.localizedString(for: start, relativeTo: Date())
    }
    
    // MARK: - Analytics Computed Properties
    
    var averageEyeContact: Double {
        let answeredQuestions = questions.compactMap { $0.answer }
        let eyeContactMetrics = answeredQuestions.compactMap { $0.eyeContactMetrics }
        guard !eyeContactMetrics.isEmpty else { return 0 }
        let total = eyeContactMetrics.reduce(0.0) { $0 + $1.percentage }
        return total / Double(eyeContactMetrics.count)
    }
    
    var averageConfidenceScore: Double {
        let answeredQuestions = questions.compactMap { $0.answer }
        let confidenceScores = answeredQuestions.compactMap { $0.confidenceScore }
        guard !confidenceScores.isEmpty else { return 0 }
        let total = confidenceScores.reduce(0.0) { $0 + $1 }
        return total / Double(confidenceScores.count)
    }
    
    var averageSpeechPace: Double {
        let answeredQuestions = questions.compactMap { $0.answer }
        let toneAnalyses = answeredQuestions.compactMap { $0.toneAnalysis }
        guard !toneAnalyses.isEmpty else { return 0 }
        let total = toneAnalyses.reduce(0.0) { $0 + $1.speechPace }
        return total / Double(toneAnalyses.count)
    }
}

// Supabase table model
struct InterviewSessionRecord: Codable {
    let id: UUID?
    let user_email: String
    let role: String
    let difficulty: String
    let categories: [String]
    let duration: Int
    let questions: String // JSON string
    let answered_count: Int
    let skipped_count: Int
    let start_time: String?
    let end_time: String?
    let created_at: String?
    
    init(from session: InterviewSession, userEmail: String) throws {
        self.id = session.id
        self.user_email = userEmail
        self.role = session.role
        self.difficulty = session.difficulty.rawValue
        self.categories = session.categories.map { $0.rawValue }
        self.duration = session.duration
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let questionsData = try encoder.encode(session.questions)
        self.questions = String(data: questionsData, encoding: .utf8) ?? "[]"
        
        self.answered_count = session.answeredCount
        self.skipped_count = session.skippedCount
        
        let dateFormatter = ISO8601DateFormatter()
        self.start_time = session.startTime.map { dateFormatter.string(from: $0) }
        self.end_time = session.endTime.map { dateFormatter.string(from: $0) }
        self.created_at = dateFormatter.string(from: Date())
    }
    
    func toSession() throws -> InterviewSession {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        
        guard let questionsData = questions.data(using: .utf8) else {
            throw NSError(domain: "InterviewModels", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid questions data"])
        }
        
        let decodedQuestions = try decoder.decode([InterviewQuestion].self, from: questionsData)
        
        let dateFormatter = ISO8601DateFormatter()
        
        return InterviewSession(
            id: id ?? UUID(),
            role: role,
            difficulty: Difficulty(rawValue: difficulty) ?? .medium,
            categories: categories.compactMap { QuestionCategory(rawValue: $0) },
            duration: duration,
            questions: decodedQuestions,
            answeredCount: answered_count,
            skippedCount: skipped_count,
            startTime: start_time.flatMap { dateFormatter.date(from: $0) },
            endTime: end_time.flatMap { dateFormatter.date(from: $0) },
            userEmail: user_email
        )
    }
}

class InterviewQuestionBank {
    static let shared = InterviewQuestionBank()
    
    let allQuestions: [InterviewQuestion] = [
        // Behavioral - Easy
        InterviewQuestion(text: "Tell me about yourself and your background.", category: .behavioral, difficulty: .easy),
        InterviewQuestion(text: "What are your greatest strengths?", category: .behavioral, difficulty: .easy),
        InterviewQuestion(text: "Why do you want to work here?", category: .behavioral, difficulty: .easy),
        InterviewQuestion(text: "Where do you see yourself in 5 years?", category: .behavioral, difficulty: .easy),
        
        // Behavioral - Medium
        InterviewQuestion(text: "Tell me about a time when you had to work under pressure.", category: .behavioral, difficulty: .medium),
        InterviewQuestion(text: "Describe a situation where you had to deal with a difficult colleague.", category: .behavioral, difficulty: .medium),
        InterviewQuestion(text: "Give me an example of when you showed leadership.", category: .behavioral, difficulty: .medium),
        InterviewQuestion(text: "Tell me about a time when you failed and what you learned from it.", category: .behavioral, difficulty: .medium),
        
        // Behavioral - Hard
        InterviewQuestion(text: "Describe a time when you had to make a difficult decision with incomplete information.", category: .behavioral, difficulty: .hard),
        InterviewQuestion(text: "Tell me about a time when you had to influence someone without authority.", category: .behavioral, difficulty: .hard),
        
        // Technical - Easy
        InterviewQuestion(text: "What programming languages are you most comfortable with?", category: .technical, difficulty: .easy),
        InterviewQuestion(text: "Explain what version control is and why it's important.", category: .technical, difficulty: .easy),
        InterviewQuestion(text: "What is your development workflow?", category: .technical, difficulty: .easy),
        
        // Technical - Medium
        InterviewQuestion(text: "Explain the difference between synchronous and asynchronous programming.", category: .technical, difficulty: .medium),
        InterviewQuestion(text: "How would you optimize a slow-running database query?", category: .technical, difficulty: .medium),
        InterviewQuestion(text: "Describe the SOLID principles and give examples.", category: .technical, difficulty: .medium),
        
        // Technical - Hard
        InterviewQuestion(text: "Design a URL shortening service like bit.ly. How would you handle scale?", category: .technical, difficulty: .hard),
        InterviewQuestion(text: "Explain how you would implement a rate limiter for an API.", category: .technical, difficulty: .hard),
        
        // Situational - Easy
        InterviewQuestion(text: "How do you prioritize your tasks when you have multiple deadlines?", category: .situational, difficulty: .easy),
        InterviewQuestion(text: "What would you do if you disagreed with a manager's decision?", category: .situational, difficulty: .easy),
        
        // Situational - Medium
        InterviewQuestion(text: "If you discovered a bug in production right before a major release, what would you do?", category: .situational, difficulty: .medium),
        InterviewQuestion(text: "How would you handle a situation where a team member isn't contributing equally?", category: .situational, difficulty: .medium),
        
        // Situational - Hard
        InterviewQuestion(text: "You notice a critical security vulnerability in a legacy system. How do you approach this?", category: .situational, difficulty: .hard),
        
        // General - Easy
        InterviewQuestion(text: "What motivates you in your work?", category: .general, difficulty: .easy),
        InterviewQuestion(text: "How do you handle feedback and criticism?", category: .general, difficulty: .easy),
        InterviewQuestion(text: "What's your ideal work environment?", category: .general, difficulty: .easy),
        
        // General - Medium
        InterviewQuestion(text: "How do you stay current with industry trends and technologies?", category: .general, difficulty: .medium),
        InterviewQuestion(text: "Describe your approach to learning new technologies.", category: .general, difficulty: .medium),
    ]
    
    func getQuestions(categories: [QuestionCategory], difficulty: Difficulty, count: Int) -> [InterviewQuestion] {
        let filtered = allQuestions.filter { question in
            categories.contains(question.category) && question.difficulty == difficulty
        }
        
        return Array(filtered.shuffled().prefix(count))
    }
}

File: FinalRoundFlutter/lib/models/interview_models.dart
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';

/// Interview question model
class InterviewQuestion {
  final String id;
  final String text;
  final QuestionCategory category;
  final Difficulty difficulty;
  QuestionAnswer? answer;

  InterviewQuestion({
    String? id,
    required this.text,
    required this.category,
    required this.difficulty,
    this.answer,
  }) : id = id ?? const Uuid().v4();

  Map<String, dynamic> toJson() => {
    'id': id,
    'text': text,
    'category': category.value,
    'difficulty': difficulty.value,
    'answer': answer?.toJson(),
  };

  factory InterviewQuestion.fromJson(Map<String, dynamic> json) {
    return InterviewQuestion(
      id: json['id'] as String?,
      text: json['text'] as String,
      category: QuestionCategory.fromString(json['category'] as String),
      difficulty: Difficulty.fromString(json['difficulty'] as String),
      answer: json['answer'] != null 
          ? QuestionAnswer.fromJson(json['answer'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// Answer to an interview question
class QuestionAnswer {
  final String transcription;
  final String? audioUrl;
  final String? videoUrl;
  final AnswerEvaluation? evaluation;
  final DateTime timestamp;
  final EyeContactMetrics? eyeContactMetrics;
  final double? confidenceScore;
  final ToneAnalysis? toneAnalysis;
  final double? timeSpent;

  QuestionAnswer({
    required this.transcription,
    this.audioUrl,
    this.videoUrl,
    this.evaluation,
    DateTime? timestamp,
    this.eyeContactMetrics,
    this.confidenceScore,
    this.toneAnalysis,
    this.timeSpent,
  }) : timestamp = timestamp ?? DateTime.now();

  Map<String, dynamic> toJson() => {
    'transcription': transcription,
    'audioUrl': audioUrl,
    'videoUrl': videoUrl,
    'evaluation': evaluation?.toJson(),
    'timestamp': timestamp.toIso8601String(),
    'eyeContactMetrics': eyeContactMetrics?.toJson(),
    'confidenceScore': confidenceScore,
    'toneAnalysis': toneAnalysis?.toJson(),
    'timeSpent': timeSpent,
  };

  factory QuestionAnswer.fromJson(Map<String, dynamic> json) {
    return QuestionAnswer(
      transcription: json['transcription'] as String,
      audioUrl: json['audioUrl'] as String?,
      videoUrl: json['videoUrl'] as String?,
      evaluation: json['evaluation'] != null
          ? AnswerEvaluation.fromJson(json['evaluation'] as Map<String, dynamic>)
          : null,
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'] as String)
          : null,
      eyeContactMetrics: json['eyeContactMetrics'] != null
          ? EyeContactMetrics.fromJson(json['eyeContactMetrics'] as Map<String, dynamic>)
          : null,
      confidenceScore: (json['confidenceScore'] as num?)?.toDouble(),
      toneAnalysis: json['toneAnalysis'] != null
          ? ToneAnalysis.fromJson(json['toneAnalysis'] as Map<String, dynamic>)
          : null,
      timeSpent: (json['timeSpent'] as num?)?.toDouble(),
    );
  }
}

/// Eye contact tracking metrics
class EyeContactMetrics {
  final double percentage; // 0-100
  final double totalDuration;
  final double lookingAtCameraDuration;
  final List<EyeContactTimestamp> timestamps;

  EyeContactMetrics({
    required this.percentage,
    required this.totalDuration,
    required this.lookingAtCameraDuration,
    this.timestamps = const [],
  });

  String get formattedPercentage => '${percentage.toStringAsFixed(1)}%';

  Map<String, dynamic> toJson() => {
    'percentage': percentage,
    'totalDuration': totalDuration,
    'lookingAtCameraDuration': lookingAtCameraDuration,
    'timestamps': timestamps.map((t) => t.toJson()).toList(),
  };

  factory EyeContactMetrics.fromJson(Map<String, dynamic> json) {
    return EyeContactMetrics(
      percentage: (json['percentage'] as num).toDouble(),
      totalDuration: (json['totalDuration'] as num).toDouble(),
      lookingAtCameraDuration: (json['lookingAtCameraDuration'] as num).toDouble(),
      timestamps: (json['timestamps'] as List<dynamic>?)
          ?.map((t) => EyeContactTimestamp.fromJson(t as Map<String, dynamic>))
          .toList() ?? [],
    );
  }
}

class EyeContactTimestamp {
  final double time;
  final bool isLookingAtCamera;

  EyeContactTimestamp({required this.time, required this.isLookingAtCamera});

  Map<String, dynamic> toJson() => {
    'time': time,
    'isLookingAtCamera': isLookingAtCamera,
  };

  factory EyeContactTimestamp.fromJson(Map<String, dynamic> json) {
    return EyeContactTimestamp(
      time: (json['time'] as num).toDouble(),
      isLookingAtCamera: json['isLookingAtCamera'] as bool,
    );
  }
}

/// Tone and speech analysis
class ToneAnalysis {
  final double speechPace; // words per minute
  final int pauseCount;
  final double averagePauseDuration;
  final double volumeVariation; // 0-1 scale
  final SentimentScore sentiment;

  ToneAnalysis({
    required this.speechPace,
    required this.pauseCount,
    required this.averagePauseDuration,
    required this.volumeVariation,
    required this.sentiment,
  });

  String get paceDescription {
    if (speechPace < 100) return 'Slow';
    if (speechPace < 140) return 'Moderate';
    if (speechPace < 180) return 'Fast';
    return 'Very Fast';
  }

  String get formattedPace => '${speechPace.toStringAsFixed(0)} WPM';

  Map<String, dynamic> toJson() => {
    'speechPace': speechPace,
    'pauseCount': pauseCount,
    'averagePauseDuration': averagePauseDuration,
    'volumeVariation': volumeVariation,
    'sentiment': sentiment.toJson(),
  };

  factory ToneAnalysis.fromJson(Map<String, dynamic> json) {
    return ToneAnalysis(
      speechPace: (json['speechPace'] as num).toDouble(),
      pauseCount: json['pauseCount'] as int,
      averagePauseDuration: (json['averagePauseDuration'] as num).toDouble(),
      volumeVariation: (json['volumeVariation'] as num).toDouble(),
      sentiment: SentimentScore.fromJson(json['sentiment'] as Map<String, dynamic>),
    );
  }
}

class SentimentScore {
  final double score; // -1 to 1
  final double confidence; // 0-1

  SentimentScore({required this.score, required this.confidence});

  String get label {
    if (score >= 0.5) return 'Very Positive';
    if (score >= 0.1) return 'Positive';
    if (score >= -0.1) return 'Neutral';
    if (score >= -0.5) return 'Negative';
    return 'Very Negative';
  }

  Map<String, dynamic> toJson() => {
    'score': score,
    'confidence': confidence,
  };

  factory SentimentScore.fromJson(Map<String, dynamic> json) {
    return SentimentScore(
      score: (json['score'] as num).toDouble(),
      confidence: (json['confidence'] as num).toDouble(),
    );
  }
}

/// Answer evaluation from AI
class AnswerEvaluation {
  final int score;
  final List<String> strengths;
  final List<String> improvements;
  final String feedback;

  AnswerEvaluation({
    required this.score,
    required this.strengths,
    required this.improvements,
    required this.feedback,
  });

  String get grade {
    if (score >= 95) return 'A+';
    if (score >= 90) return 'A';
    if (score >= 85) return 'A-';
    if (score >= 80) return 'B+';
    if (score >= 75) return 'B';
    if (score >= 70) return 'B-';
    if (score >= 65) return 'C+';
    if (score >= 60) return 'C';
    if (score >= 55) return 'C-';
    return 'D';
  }

  Color get gradeColor {
    if (score >= 90) return Colors.green;
    if (score >= 80) return Colors.blue;
    if (score >= 70) return Colors.orange;
    return Colors.red;
  }

  Map<String, dynamic> toJson() => {
    'score': score,
    'strengths': strengths,
    'improvements': improvements,
    'feedback': feedback,
  };

  factory AnswerEvaluation.fromJson(Map<String, dynamic> json) {
    return AnswerEvaluation(
      score: json['score'] as int,
      strengths: (json['strengths'] as List<dynamic>).cast<String>(),
      improvements: (json['improvements'] as List<dynamic>).cast<String>(),
      feedback: json['feedback'] as String,
    );
  }
}

/// Question category enum
enum QuestionCategory {
  behavioral('Behavioral'),
  technical('Technical'),
  situational('Situational'),
  general('General');

  final String value;
  const QuestionCategory(this.value);

  static QuestionCategory fromString(String value) {
    return QuestionCategory.values.firstWhere(
      (e) => e.value == value,
      orElse: () => QuestionCategory.general,
    );
  }
}

/// Difficulty level enum
enum Difficulty {
  easy('Easy'),
  medium('Medium'),
  hard('Hard');

  final String value;
  const Difficulty(this.value);

  static Difficulty fromString(String value) {
    return Difficulty.values.firstWhere(
      (e) => e.value == value,
      orElse: () => Difficulty.medium,
    );
  }
}

/// Interview session model
class InterviewSession {
  final String id;
  final String role;
  final Difficulty difficulty;
  final List<QuestionCategory> categories;
  int duration; // in seconds
  List<InterviewQuestion> questions;
  int answeredCount;
  int skippedCount;
  DateTime? startTime;
  DateTime? endTime;
  String? userEmail;
  bool enableAudioRecording;
  String experienceLevel;

  InterviewSession({
    String? id,
    required this.role,
    required this.difficulty,
    required this.categories,
    this.duration = 0,
    required this.questions,
    this.answeredCount = 0,
    this.skippedCount = 0,
    this.startTime,
    this.endTime,
    this.userEmail,
    this.enableAudioRecording = true,
    this.experienceLevel = 'Mid Level',
  }) : id = id ?? const Uuid().v4();

  int get totalQuestions => questions.length;
  int get attemptedQuestions => answeredCount + skippedCount;
  int get remainingQuestions => totalQuestions - attemptedQuestions;

  Duration? get sessionDuration {
    if (startTime == null || endTime == null) return null;
    return endTime!.difference(startTime!);
  }

  double get completionRate {
    if (totalQuestions == 0) return 0;
    return answeredCount / totalQuestions;
  }

  double get averageScore {
    final evaluatedQuestions = questions
        .where((q) => q.answer?.evaluation != null)
        .map((q) => q.answer!.evaluation!)
        .toList();
    if (evaluatedQuestions.isEmpty) return 0;
    final totalScore = evaluatedQuestions.fold<int>(0, (sum, e) => sum + e.score);
    return totalScore / evaluatedQuestions.length;
  }

  String get overallGrade {
    final score = averageScore.round();
    if (score >= 95) return 'A+';
    if (score >= 90) return 'A';
    if (score >= 85) return 'A-';
    if (score >= 80) return 'B+';
    if (score >= 75) return 'B';
    if (score >= 70) return 'B-';
    if (score >= 65) return 'C+';
    if (score >= 60) return 'C';
    if (score >= 55) return 'C-';
    return 'D';
  }

  Color get gradeColor {
    final score = averageScore.round();
    if (score >= 90) return Colors.green;
    if (score >= 80) return Colors.blue;
    if (score >= 70) return Colors.orange;
    return Colors.red;
  }

  double get answerRate {
    if (attemptedQuestions == 0) return 0;
    return answeredCount / attemptedQuestions;
  }

  String get formattedDuration {
    final dur = sessionDuration;
    if (dur == null) return 'N/A';
    final minutes = dur.inMinutes;
    final seconds = dur.inSeconds % 60;
    return '${minutes}m ${seconds}s';
  }

  String get formattedDate {
    final date = startTime ?? endTime ?? DateTime.now();
    final now = DateTime.now();
    final diff = now.difference(date);
    
    if (diff.inDays == 0) {
      return 'Today';
    } else if (diff.inDays == 1) {
      return 'Yesterday';
    } else if (diff.inDays < 7) {
      return '${diff.inDays} days ago';
    } else {
      return '${date.day}/${date.month}/${date.year}';
    }
  }

  double get averageEyeContact {
    final metrics = questions
        .where((q) => q.answer?.eyeContactMetrics != null)
        .map((q) => q.answer!.eyeContactMetrics!)
        .toList();
    if (metrics.isEmpty) return 0;
    return metrics.fold<double>(0, (sum, m) => sum + m.percentage) / metrics.length;
  }

  double get averageConfidenceScore {
    final scores = questions
        .where((q) => q.answer?.confidenceScore != null)
        .map((q) => q.answer!.confidenceScore!)
        .toList();
    if (scores.isEmpty) return 0;
    return scores.fold<double>(0, (sum, s) => sum + s) / scores.length;
  }

  double get averageSpeechPace {
    final analyses = questions
        .where((q) => q.answer?.toneAnalysis != null)
        .map((q) => q.answer!.toneAnalysis!)
        .toList();
    if (analyses.isEmpty) return 0;
    return analyses.fold<double>(0, (sum, a) => sum + a.speechPace) / analyses.length;
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'role': role,
    'difficulty': difficulty.value,
    'categories': categories.map((c) => c.value).toList(),
    'duration': duration,
    'questions': questions.map((q) => q.toJson()).toList(),
    'answeredCount': answeredCount,
    'skippedCount': skippedCount,
    'startTime': startTime?.toIso8601String(),
    'endTime': endTime?.toIso8601String(),
    'userEmail': userEmail,
    'enableAudioRecording': enableAudioRecording,
    'experienceLevel': experienceLevel,
  };

  factory InterviewSession.fromJson(Map<String, dynamic> json) {
    return InterviewSession(
      id: json['id'] as String?,
      role: json['role'] as String,
      difficulty: Difficulty.fromString(json['difficulty'] as String),
      categories: (json['categories'] as List<dynamic>)
          .map((c) => QuestionCategory.fromString(c as String))
          .toList(),
      duration: json['duration'] as int? ?? 0,
      questions: (json['questions'] as List<dynamic>)
          .map((q) => InterviewQuestion.fromJson(q as Map<String, dynamic>))
          .toList(),
      answeredCount: json['answeredCount'] as int? ?? 0,
      skippedCount: json['skippedCount'] as int? ?? 0,
      startTime: json['startTime'] != null
          ? DateTime.parse(json['startTime'] as String)
          : null,
      endTime: json['endTime'] != null
          ? DateTime.parse(json['endTime'] as String)
          : null,
      userEmail: json['userEmail'] as String?,
      enableAudioRecording: json['enableAudioRecording'] as bool? ?? true,
      experienceLevel: json['experienceLevel'] as String? ?? 'Mid Level',
    );
  }
}
